### PR TITLE
RsXembly can modify  DOM documents

### PR DESCRIPTION
**Pull request by Happy-Neko**
Fix for #462: `RsXembly` can modify  DOM documents.

### CODE REVIEW

**Comment by Happy-Neko** [comment]
@davvd Please proceed with code review.

**Comment by davvd** [comment]
@happy-neko Thanks, let me find someone who can review this pull request

**Comment by davvd** [comment]
@dmzaytsev review this one,please

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by yegor256** [comment]
@Happy-Neko I see one big problem here. You will modify the `Document`. Not just generate an XML page, but also modify the document provided to you. This side-effect is really not what I, as a user, would expect. See the problem? I would suggest to create a unit test, proving that I'm right (or maybe I'm wrong) and fixing the code.

**Comment by Happy-Neko** [comment]
@yegor256 Fix committed.

**Comment by yegor256** [comment]
@Happy-Neko I'm curious, why not just `cloneNode(true)`?

**Comment by Happy-Neko** [comment]
@yegor256 
I don't have first hand experience but people on [stackoverflow](https://stackoverflow.com/questions/5226852/cloning-dom-document-object#comment31820298_5227450) suggest `cloneNode(true)` don't do exact copy.

**Comment by Happy-Neko** [comment]
@yegor256 Please confirm PR merge.

File: src/main/java/org/takes/rs/xe/RsXembly.java [hunk 1]
```diff
@@ -120,4 +156,37 @@ private static InputStream render(final XeSource src) throws IOException {
...
+     * Create empty DOM Document.
+     * @return Document
+     */
+    private static Document emptyDocument() {
+        try {
+            return DocumentBuilderFactory.newInstance()
+                .newDocumentBuilder()
+                .newDocument();
+        } catch (final ParserConfigurationException ex) {
+            throw new IllegalStateException(ex);
```
Comment (yegor256) [inline src/main/java/org/takes/rs/xe/RsXembly.java]: @Happy-Neko see http://www.yegor256.com/2015/12/01/rethrow-exceptions.html

File: src/main/java/org/takes/rs/xe/RsXembly.java [hunk 2]
```diff
@@ -120,4 +156,37 @@ private static InputStream render(final XeSource src) throws IOException {
...
+    private static Node cloneNode(final Node dom) {
+        try {
+            final Transformer transformer =
+                TransformerFactory.newInstance().newTransformer();
+            final DOMSource source = new DOMSource(dom);
+            final DOMResult result = new DOMResult();
+            transformer.transform(source, result);
+            return result.getNode();
+        } catch (final TransformerConfigurationException ex) {
+            throw new IllegalStateException(ex);
```
Comment (yegor256) [inline src/main/java/org/takes/rs/xe/RsXembly.java]: @Happy-Neko same here, see http://www.yegor256.com/2015/12/01/rethrow-exceptions.html

File: src/main/java/org/takes/rs/xe/RsXembly.java [hunk 3]
```diff
@@ -120,4 +156,37 @@ private static InputStream render(final XeSource src) throws IOException {
...
+        } catch (final TransformerException ex) {
+            throw new IllegalArgumentException(ex);
```
Comment (yegor256) [inline src/main/java/org/takes/rs/xe/RsXembly.java]: @Happy-Neko and here

**Comment by yegor256** [comment]
@Happy-Neko I didn't know that... good to know. I made a few small comments above

**Comment by Happy-Neko** [comment]
@yegor256 changes committed. Thanks.

**Comment by yegor256** [comment]
@rultor merge

**Comment by davvd** [comment]
@elenavolokhova review this ticket please, for compliance with our [quality rules](http://at.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor deploy

**Comment by elenavolokhova** [comment]
@dmzaytsev 
According to our [quality rules](http://www.teamed.io/qa.html):
 > The code reviewer found at least three problems in the code.
 
 There were no issues found during review.
Please confirm that you will try to find at least three problems next time.

**Comment by dmzaytsev** [comment]
@elenavolokhova I confirm

**Comment by elenavolokhova** [comment]
@davvd Quality is acceptable here.

**Comment by davvd** [comment]
> @davvd Quality is acceptable here.

@elenavolokhova thank you, let's try to make it "good" next time

**Comment by davvd** [comment]
@dmzaytsev paied 10 mins to @elenavolokhova for QA review (payment ID is `74073171`); **17 mins** was added to your account, many thanks for your contribution (payment 74073174)! The task took 364 hours and 57 mins.; you're getting extra minutes here (c=2); added +17 to your rating, now it is equal to [+1524](http://www.netbout.com/b/35166?open=rating)

**Pull request merged**